### PR TITLE
Add support for timeout option

### PR DESCRIPTION
**Pull request by davidkudera**
Based on #328 and #411.

Unfortunately it seems that php-ssh doesn't support timeout - Herzult/php-ssh#46

### CODE REVIEW

**Comment by elfet** [comment]
Cool! Need to code review and test it. 
@deployphp/maintainer please, help!

**Comment by oanhnn** [comment]
Cool!
I think it is better if support:
```
localServer($name, $timeout = 300);
// and
server($name, $host, $port = 22, $timeout =300);
```

**Comment by davidkudera** [comment]
@oanhnn Well, I made it like this because @elfet said that it would be better (https://github.com/deployphp/deployer/pull/328#discussion_r32095173). If I understand correctly

**Comment by oanhnn** [comment]
@elfet What do you think?
@davidkudera Can you write docs and create a PR to https://github.com/deployphp/docs

**Comment by davidkudera** [comment]
@oanhnn sure, I'll do that right after you'll decide which version is better (method or argument)

**Comment by oanhnn** [comment]
@elfet Do you decide which version is better (method or argument).
```
server(...)
    ->timeout(300)
```
or 
```
server('dev', 192.168.1.2, 22, 300)
```
or both :smile:

**Comment by elfet** [comment]
I dont think what is good to setup this server wide. Its better to specify timeout on run:
```php
run(‘…’, 360);
```

**Comment by davidkudera** [comment]
@elfet yes that is only logical since `runLocally` also has it. Actually I was thinking about it too, but I didn't want to do that in my first PR because it worked just for local server and it would mean to change implementation of `ServerInterface`. Also I made just the changes mentioned in the second PR. 

But what if we'll change `ServerInterface::run()` but keep global server configuration too? If user provide second argument at `run` method, we'll use that, othervise we'll use value from server configuration. And for people using ssh we can throw some `NotImplementedException` for example.

What do you think?

**Comment by elfet** [comment]
I think global config here is bad decision.

**Comment by davidkudera** [comment]
Ok, so should I change it? @elfet @oanhnn

**Comment by elfet** [comment]
I think yes :)

**Comment by davidkudera** [comment]
@elfet ok, I'll force push these changes. Unfortunately right now I don't have time, but I should be back home maybe at sunday

**Comment by davidkudera** [comment]
@elfet @oanhnn what do you think now guys?

File: src/Server/Remote/PhpSecLib.php [hunk 1]
```diff
@@ -102,10 +104,15 @@ public function checkConnection()
...
+    public function run($command, $timeout = null)
     {
+        if ($timeout === null) {
+            $timeout = self::TIMEOUT;
+        }
+
         $this->checkConnection();
 
+        $this->sftp->setTimeout($timeout);
         $result = $this->sftp->exec($command);
```
Comment (torinaki) [inline src/Server/Remote/PhpSecLib.php]: How about to add here check that operation wasn't reached timeout like that
        
        if ($this->sftp->isTimeout()) {
            throw new \RuntimeException('Operation timeout');
        }
        
// cc @elfet , @oanhnn

**Comment by davidkudera** [comment]
ping @elfet @oanhnn

**Comment by davidkudera** [comment]
rebased to master

File: src/Server/Remote/PhpSecLib.php (hunk 1 above)
Comment (elfet) [inline src/Server/Remote/PhpSecLib.php]: What is `isTimeout()`?

**Comment by elfet** [comment]
```
Timeout is not supported in ssh connection.
```
WE really need to figure out how to implement this in ssh2 ext. Otherwise timeout with be really strange functionality.

File: src/Server/Remote/PhpSecLib.php (hunk 1 above)
Comment (torinaki) [inline src/Server/Remote/PhpSecLib.php]: @elfet it is related to my latest temp.fixed issue with timeout(see PR #429 )
Currenlty if timeout will be reached deployer will silently stop task without any notification and will continue with the next one. This exception will at least help to understand what was happen. 

This flag will be set to true if timeout happen:
https://github.com/phpseclib/phpseclib/blob/7da4f2e6307d11c5984906dafe3c53d4f7c157a1/phpseclib/Net/SSH2.php#L2823

Here is oficial documentation about timeout: http://phpseclib.sourceforge.net/ssh/examples.html#timeout

Comment (torinaki) [inline src/Server/Remote/PhpSecLib.php]: BTW `$this->sftp->exec($command)` will not throw any exception about timeout

**Comment by davidkudera** [comment]
Yeah, but do we know if ssh2 is still developed? Last commit was 5 months ago..

**Comment by torinaki** [comment]
@elfet , @davidkudera SSH wrapper library use stream blocking. It's easy to refactor code to imlement not blocking implementation with callback:
https://github.com/Herzult/php-ssh/blob/d569d15483a5f9218700313cd21c50090aa93351/src/Ssh/Exec.php#L26

But I'm still wonder about use cases of this timeout. @davidkudera can you please provide some examples of use cases? Specialy for me. I'm just curious.

**Comment by torinaki** [comment]
Or another idea is to wrap command with some shell timeout function.
Windows: http://stackoverflow.com/questions/12394338/windows-batch-analogue-for-timeout-command
Linux: http://man7.org/linux/man-pages/man1/timeout.1.html
Without any magic :)

**Comment by davidkudera** [comment]
@torinaki sure, for example we were developing new version of our website for more than half of a year (without touching 1st version). My deploy process failed almost immediatelly - to be more specific at database migrations. On my PC with linux it runs for about 15 minutes and I wasn't even brave enough to try it on Windows :grinning:

**Comment by torinaki** [comment]
@davidkudera for what exact task you suppose to use `timeout`? Database migrations?

**Comment by davidkudera** [comment]
Yes, for now only for database migrations, but maybe there will be some other tasks too.

File: src/Server/Remote/SshExtension.php [hunk 2]
```diff
@@ -109,8 +109,12 @@ public function checkConnection()
     /**
      * {@inheritdoc}
      */
-    public function run($command)
+    public function run($command, $timeout = null)
     {
+        if ($timeout !== null) {
+            throw new \RuntimeException('Timeout is not supported in ssh connection.');
```
Comment (elfet) [inline src/Server/Remote/SshExtension.php]: So sad( But if some one try to run it with ext-ssh?

Comment (davidkudera) [inline src/Server/Remote/SshExtension.php]: It should just throw this exception, or what do you mean? Maybe the exception should have better message?

**Comment by davidkudera** [comment]
I just noticed that this PR is still open. Is there anything I can do now to make it merged?

The problem with SSH is unfortunate but I think that the solution proposed in this PR is still much better than the current state of deployer.

**Comment by elfet** [comment]
Ok, i'll test this PR to merge. Need some time.

**Comment by ceesvanegmond** [comment]
Still open, @elfet When does it get merged?

**Comment by torinaki** [comment]
@elfet from my point of view it is important that following comment should be addressed before merge:
https://github.com/deployphp/deployer/pull/415#discussion_r37126085
otherwise deployer will silently stop task in case when it wil reaches timeout.

**Comment by TheRatG** [comment]
+1

**Pull request closed without merging**